common/formats/cue.cpp

```diff
@@ -312,18 +314,62 @@ void CueSheet::parseTracksContext(const char *line) {
 			flags |= lookupInTable(trackFlags, flag.c_str());
 		}
 
-		_files[_currentFile].tracks[_currentTrack].flags = flags;
+		_tracks[_currentTrack].flags = flags;
 
-		debug(5, "Track flags: %d", _files[_currentFile].tracks[_currentTrack].flags);
+		debug(5, "Track flags: %d", _tracks[_currentTrack].flags);
 	} else if (command == "FILE") {
 		parseHeaderContext(line);
 	} else if (command == "PERFORMER") {
-		_files[_currentFile].tracks[_currentTrack].performer = nexttok(s, &s);
+		_tracks[_currentTrack].performer = nexttok(s, &s);
 
-		debug(5, "Track performer: %s", _files[_currentFile].tracks[_currentTrack].performer.c_str());
+		debug(5, "Track performer: %s", _tracks[_currentTrack].performer.c_str());
 	} else {
 		warning("CueSheet: Unprocessed track command %s at line %d", command.c_str(), _lineNum);
 	}
 }
 
+Array<CueSheet::CueFile> CueSheet::files() {
+	return _files;
+}
+
+Array<CueSheet::CueTrack> CueSheet::tracks() {
+	return _tracks;
+}
+
+CueSheet::CueTrack *CueSheet::getTrack(int tracknum) {
+	for (int i = 0; i < _tracks.size(); i++) {
+		if (_tracks[i].number == tracknum) {
+			return &_tracks[i];
+		}
+	}
+
+	return nullptr;
+}
+
+CueSheet::CueTrack *CueSheet::getTrackAtFrame(int frame) {
+	for (int i = 0; i < _tracks.size(); i++) {
+		// Inside pregap
+		if (frame >= _tracks[i].indices[0] && frame < _tracks[i].indices.back()) {
+			debug(5, "CueSheet::getTrackAtFrame: Returning track %i (pregap)", i);
+			return &_tracks[i];
+		}
+
+		// Between index 1 and the start of the subsequent track
+		if (i < _tracks.size() && frame > _tracks[i].indices.back() && frame < _tracks[i+1].indices[0]) {
+			debug(5, "CueSheet::getTrackAtFrame: Returning track %i (inside content)", i);
+			return &_tracks[i];
+		}
+	}
+
+	// Not found within any tracks, but could be in the final track
+	if (frame > _tracks.back().indices.back()) {
+		debug(5, "CueSheet::getTrackAtFrame: Returning final track");
```

Comment:
I would turn it into a `warning()`, e.g. the user code is requesting us to do something odd, beyond the CD limits.

Reply:
The request isn't as odd as it sounds - I'll update the comment to explain why.

The cuesheet gives us the *starting* indices of each track, but not track length or ending indices. In other words, we don't know how long the song is - we actually don't have an indication of whether the index the caller passed is in the disc or not, we just know that it comes after the start of this track.

The only actual way to 100% for sure know the number of sectors in the track is to actually check the file on disk and see how large it is. (Or load it as an audio file and check the duration.) That's a bit too complex to handle here; I'd rather defer complexity around loading and examining audio files to the CD audio player.